File: solutions/124_word_split.go
package main

// dp[i]表示s的前i个字符是否匹配
func wordBreak(s string, wordDict []string) bool {
	// 使用哈希表 可以快速判断是否出现过此字符串
	hashmap := map[string]bool{}
	for _, word := range wordDict {
		hashmap[word] = true
	}
	n := len(s)
	dp := make([]bool, n+1)
	dp[0] = true // 空字符串为true
	for i := 1; i <= n; i++ {
		for j := 0; j < i; j++ {
			dp[i] = dp[i] || dp[j] && hashmap[s[j:i]]
		}
	}
	return dp[n]
}

File: solutions/__124_word_split.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

public class __124_word_split {
    public static void main(String[] args) {
        __124_word_split solution = new __124_word_split();
        System.out.println(solution.wordBreak("leetcode", Arrays.asList("leet", "code")));
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        HashSet<String> set = new HashSet<>(wordDict);
        int n = s.length();
        boolean[] dp = new boolean[n+1];
        dp[0] = true;
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                dp[i] = dp[j] && set.contains(s.substring(j,i));
                if (dp[i]) {
                    break;
                }
            }
        }
        return dp[n];
    }
}

